### ISSUE TITLE
Change BigFloat exponent range dynamically/on startup?

### ISSUE THREAD

**Issue by octonion** [JuliaLang/julia#9816]
Looks like an exponent issue.

```julia
julia> x = BigFloat(2)
2e+00 with 256 bits of precision

julia> y = x^(2^29.999)
2.147630010985401604264819292317090866477916691008148747595402140079551131402097e+323004529 with 256 bits of precision

julia> y = x^(2^30)
inf with 256 bits of precision
```

**Comment by jakebolewski** [comment]
You can set the precision with `with_bigfloat_precision` or `set_bigfloat_precision`.

```julia
y = with_bigfloat_precision(1024) do
       x = BigFloat(2)
       x^(2^29.999)
end
2.147630010985401604264819292317090866477916691008148747595402140079551131402103915021620139582322551521799422736825697062432910266153373554562651896535013449148123040212518737728306124908676242374509907454472106168350342712144311640608331536812521116427091127566166578972087298063550247239907705892487661701698e+323004529 with 1024 bits of precision
``` 
( I just made the # bits of precision up)

Is this not clear or discoverable from the manual?  Please let us know how we might be able to improve documenting this!

**Comment by jakebolewski** [comment]
Whoops, sorry I misread your example.

**Comment by octonion** [comment]
I'm cloning to see if I can identify the issue and whether or not it's easily fixable.

**Comment by jiahao** [comment]
Taking the limit from below, this looks like correct behavior.

```jl
julia> prevfloat(typemax(BigFloat)) #Largest finite BigFloat in current precision
2.09857871646738769240435811688383907063809796547335262778664622571024044777575e+323228496 with 256 bits of precision

julia> BigFloat(2)^float(2^29.999999)
3.785398101107661455980339451987360243961883630030321525306082346663052761621028e+323228272 with 256 bits of precision

julia> BigFloat(2)^float(2^29.9999999)
1.653726479041330888683573563567023467201627435451942689785951702393846104232551e+323228474 with 256 bits of precision

julia> BigFloat(2)^float(2^29.99999999)
2.412714074650245865034617987469335714877507459384761606388480546546563111360727e+323228494 with 256 bits of precision

julia> BigFloat(2)^float(2^29.999999999)
inf with 256 bits of precision
```

**Comment by octonion** [comment]
A 30 bit exponent? 31 bits with sign?

**Comment by jiahao** [comment]
I'm not sure what you are trying to say. The exponents of the form `2^29.99...` are `Float64`s.

All that is happening is that your computation is overflowing beyond the largest representable floating point number in the current precision.

Another way to see this is to take logs.

```jl
julia> expmax = log2(prevfloat(typemax(BigFloat)))
1.073741823e+09 with 256 bits of precision

julia> expmax - 2^29.99999999
6.44261181354522705078125e+00 with 256 bits of precision

julia> expmax - 2^29.999999999 
-2.5573885440826416015625e-01 with 256 bits of precision
```

**Comment by octonion** [comment]
I'm not sure what you're saying. Is the below the expected behavior?

```julia
julia> set_bigfloat_precision(10000)
10000

julia> x = BigFloat(2)
2e+00 with 10000 bits of precision

julia> y = BigFloat(2^30)
1.073741824e+09 with 10000 bits of precision

julia> x^y
inf with 10000 bits of precision
```

**Comment by octonion** [comment]
My apologies; looks like that's the limit for BigFloat. Sage seems to use a 64-bit exponent.

```python
sage: x = 2.0
sage: y = 2.0**61
sage: x**y
3.42801802478096e694127911065419641
```

**Comment by jiahao** [comment]
As far as I can tell, the exponent range in MPFR is platform dependent.

**Comment by jakebolewski** [comment]
What bigfloat library does Sage use?  Does it use Python's bigfloat support by default?

**Comment by octonion** [comment]
You're right, Julia uses mpfr. E_min and E_max are settable in mpfr, but Julia seem to be using the defaults of E_min = 1-2^30 and  E_max = 2^30-1.

I came across this due to some overflow in my Pell equation solver.

**Comment by jiahao** [comment]
I believe SAGE uses Maxima's bigfloat library under the hood, not MPFR.

Ref: [thread](http://sourceforge.net/p/maxima/mailman/maxima-discuss/thread/5303CF0C.7020406@berkeley.edu/)

**Comment by octonion** [comment]
Just checked my system; for some reason Julia is defaulting to less than the max allowed exponent range.

```python
>>> import gmpy2
>>> from gmpy2 import mpfr
>>> gmpy2.get_emax_max()
4611686018427387903
```

4611686018427387903 = 2^62-1

**Comment by jakebolewski** [comment]
I was just looking at the `gmpy` webpage :-)  There is no reason why we cannot have this functionality, it is just that no one has exposed it yet.  @octonion it would make a great pull request.

**Comment by jiahao** [comment]
Looks like all that is really needed is to `set_emax` to its maximal allowable default:

```jl
julia> ccall( (:mpfr_get_emax, :libmpfr), Culong, ())
0x000000003fffffff

julia> ccall( (:mpfr_set_emax, :libmpfr), Void, (Culong,), 2^62-1)

julia> BigFloat(2)^(2^62-2)
2.937826894555793795468455999439221294969258196372749154166889803234661792194963e+1388255822130839282 with 256 bits of precision
```

Reopening as a question about whether we want to initialize this on startup.

**Comment by jakebolewski** [comment]
So it defaults to 30 because it is least common denominator exponent between 32 and 64 bit platforms (and hence the default is portable)?

We don't usually hide platform differences if that is the case.

**Comment by octonion** [comment]
Thanks, Jiahao! It's great that you can do this within Julia.

**Comment by octonion** [comment]
Everything works as expected with mpfr's mexp set to 2^62-1.

```julia
squarefree equation: 4729494*x^2 + 1 = y^2
squarefree solution = [50549485234315033074477819735540408986340, 109931986732829734979866232821433543901088049, 1]
steps = 91
power = 140802024
full equation: 4570862300998510050787909656*x^2 + 1 = y^2
smallest (x,y) have (6243465028,6243465042) digits
```

Code is here: https://github.com/octonion/puzzles/blob/master/chakravala/bhaskara_squarefree_squaring.jl

Example was called with the argument 4570862300998510050787909656 after
`ccall( (:mpfr_set_emax, :libmpfr), Void, (Clong,), 2^62-1)`
was added at the top. Takes about 1 minute to calculate the fundamental squarefree solution and to calculate the needed power, then another 45 minutes to calculate the fundamental solution of the full equation via exponentiation by squaring. The smallest (x,y) have over 6 billion digits each.

**Comment by StefanKarpinski** [comment]
This seems like a case where using 32 bits on 32-bit platforms is pointless. This stuff is going to be relatively slow regardless and we may as well set the max range by default.

**Comment by jakebolewski** [comment]
Closed by #9827.
